presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/DatabaseMetadataWriter.java

```diff
@@ -0,0 +1,728 @@
+            if (commit == null) {
+                return null;
+            }
+            if (!commit.isRollingBack()) {
+                dao.abortActiveCommit();
+            }
+
+            // rollback master
+            dao.rollback(commit.getCommitId());
+
+            return commit;
+        });
+
+        // skip if no rollback needed
+        if (activeCommit == null) {
+            return;
+        }
+
+        long commitId = activeCommit.getCommitId();
+        RollbackInfo rollbackInfo = activeCommit.getRollbackInfo();
+
+        // rollback shards
+        for (Jdbi shardDbi : shardDbi) {
+            shardDbi.useTransaction(handle -> {
+                ShardWriterDao dao = handle.attach(ShardWriterDao.class);
+
+                // acquire shard lock
+                long abortedCommitId = getLockedAbortedCommitId(dao);
+
+                // skip if no rollback needed
+                if (abortedCommitId > commitId) {
+                    return;
+                }
+
+                dao.updateAbortedCommitId(commitId);
+
+                // rollback shard
+                dao.rollback(commitId);
+
+                for (long tableId : rollbackInfo.getWrittenTableIds()) {
+                    dao.rollbackCreatedIndexChunks(chunkIndexTable(tableId), commitId);
+                    dao.rollbackDeletedIndexChunks(chunkIndexTable(tableId), commitId);
+                }
+            });
+        }
+
+        // rollback indexes
+        for (long tableId : rollbackInfo.getCreatedTableIds()) {
+            shardDbi.get(getTableShard(tableId)).useHandle(handle ->
+                    dropIndexTable(handle, tableId));
+        }
+
+        for (TableColumn column : rollbackInfo.getAddedColumns()) {
+            shardDbi.get(getTableShard(column.getTableId())).useHandle(handle ->
+                    dropIndexTableColumn(databaseType, handle, column.getTableId(), column.getColumnId()));
+        }
+
+        // finish rollback
+        masterDbi.useTransaction(handle -> {
+            MasterWriterDao dao = handle.attach(MasterWriterDao.class);
+
+            // acquire global lock
+            getLockedCurrentCommitId(dao);
+
+            // verify we are still rolling back
+            ActiveCommit commit = dao.getActiveCommit();
+            if ((commit == null) || (commit.getCommitId() != commitId)) {
+                return;
+            }
+            verifyMetadata(commit.isRollingBack(), "Commit should be rolling back");
+
+            // finalize rollback
+            deleteActiveCommit(dao);
+        });
+    }
+
+    @Override
+    public long beginCommit()
+    {
+        return masterDbi.inTransaction(handle -> {
+            MasterWriterDao dao = handle.attach(MasterWriterDao.class);
+
+            // acquire global lock
+            long currentCommitId = getLockedCurrentCommitId(dao);
+
+            // verify no commit in progress
+            if (dao.getActiveCommit() != null) {
```

Comment:
I understand this is for the rare case that we run 2 coordinates, and we still want only 1 commit at a time.

But the following can happen, I think:
t1 runs 2 hours
t2 commit and throw exception, then activeCommit is t2
t1 commit, fails because activeCommit
t2 rollback
t1 rollback
Before a failure got chance to rollback, everything will fail. It's not logical error. But is it fair design? just fail for irrelevant things?

Reply:
Sorry, I just read again. This cannot happen in current logic. 

"t1 commit, fails because activeCommit"  cannot happen.

t1 commit, need to recover
activeCommit == t2 is rolledback
t1 commit succeeds

Very good logic. Sorry if waste your time.